opencsp/app/sofast/lib/ProcessSofastAbstract.py

```diff
@@ -0,0 +1,296 @@
+        General image processing data for the calculation.
+    data_geometry_facet : list[cdc.CalculationDataGeometryFacet]
+        List of geometry data for each facet.
+    data_image_processing_facet : list[cdc.CalculationImageProcessingFacet]
+        List of image processing data for each facet.
+    data_error : cdc.CalculationError
+        Error data for the calculation.
+    data_calculation_facet : list[SlopeSolverData]
+        List of slope solver data for each facet.
+    data_calculation_ensemble : list[cdc.CalculationFacetEnsemble]
+        List of calculation data for the ensemble.
+    params : HDF5_SaveAbstract
+        Parameters for the calculation.
+
+    Methods
+    -------
+    save_to_hdf(file: str, prefix: str = '')
+        Saves data to the given HDF file.
+    get_optic(interp_type: Literal['bilinear', 'clough_tocher', 'nearest'] = 'nearest') -> FacetEnsemble | Facet
+        Returns the OpenCSP representation of the optic under test.
+    """
+
+    def __init__(self):
+        self.num_facets: int = None
+        self.optic_type: Literal['undefined', 'single', 'multi'] = None
+        self.data_facet_def: list[DefinitionFacet] = None
+        self.data_ensemble_def: DefinitionEnsemble = None
+
+        self.data_surfaces: list[Surface2DAbstract] = None
+
+        self.data_geometry_general: cdc.CalculationDataGeometryGeneral = None
+        self.data_image_processing_general: cdc.CalculationImageProcessingGeneral = None
+        self.data_geometry_facet: list[cdc.CalculationDataGeometryFacet] = None
+        self.data_image_processing_facet: list[cdc.CalculationImageProcessingFacet] = None
+        self.data_error: cdc.CalculationError = None
+
+        self.data_calculation_facet: list[SlopeSolverData] = None
+        self.data_calculation_ensemble: list[cdc.CalculationFacetEnsemble] = None
+
+        self.params: HDF5_SaveAbstract = None
+
+    def save_to_hdf(self, file: str, prefix: str = ''):
+        """Saves data to given file. Data is stored as: PREFIX + Folder/Field_1
+
+        Parameters
+        ----------
+        file : str
+            HDF file to save to
+        prefix : str
+            Prefix to append to folder path within HDF file (folders must be separated by "/")
+        """
+        # Log
+        lt.info(f'Saving Sofast data to: {file:s}, in HDF5 folder: "{prefix:s}"')
+
+        # One per measurement
+        if self.data_error is not None:
+            self.data_error.save_to_hdf(file, f'{prefix:s}DataSofastCalculation/general/')
+        self.data_geometry_general.save_to_hdf(file, f'{prefix:s}DataSofastCalculation/general/')
+        self.data_image_processing_general.save_to_hdf(file, f'{prefix:s}DataSofastCalculation/general/')
+
+        # Sofast parameters
+        self.params.save_to_hdf(file, f'{prefix:s}DataSofastInput/')
+
+        # Facet definition
+        if self.data_facet_def is not None:
+            for idx_facet, facet_data in enumerate(self.data_facet_def):
```

Comment:
is enumerate guaranteed to not return a integer value greater than 999?

Reply:
Technically no. But practically yes; that will likely never happen. idx_facet is the index of the facet being measured in a multi-facet array. I was trying to balance simplicity (always having a three digit number) with consistency (the chances someone will have >999 facets is very small). 

Do you think this reasoning is alright?

Reply:
If you anticipate any bugs related to values larger than 999, I would strongly suggest adding a bounds check where appropriate to ensure that that largest facet index does not exceed 999, and, if it does, throw an exception stating that this is not permitted. I confirmed that the width specifier does not truncate values larger than 999 in the f-strings, so that should be fine.